File: Problem44.java
//Time Complexity: O(n^2)
//Space Complexity: O(n*n)with recursive stack - creating array size n for each tree node
//Did it run on leetcode: yes
//Problems faced any: No

import java.util.Arrays;

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode() {}
    TreeNode(int val) { this.val = val; }
    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

public class Problem44 {
    public TreeNode buildTree(int[] inorder, int[] postorder) {
        if(postorder.length==0 || inorder.length==0)
            return null;

        //root is the last element in postorder 
        TreeNode root = new TreeNode(postorder[postorder.length-1]);
        int index = -1;

        //finding root in inorder
        for(int i=0; i<inorder.length; i++){
            if(inorder[i]==root.val)
                index = i;
        }



        //finding left part in both inorder and postorder
        int[] leftInorder = Arrays.copyOfRange(inorder, 0, index);
        int[] leftPostorder = Arrays.copyOfRange(postorder, 0, index);
        //finding right part in both inorder and postorder
        int[] rightInorder = Arrays.copyOfRange(inorder, index+1, inorder.length);
        int[] rightPostorder = Arrays.copyOfRange(postorder, index, postorder.length-1);


        root.left = buildTree(leftInorder, leftPostorder);
        root.right = buildTree(rightInorder, rightPostorder);

        return root;

    }
}

File: Problem-44.py
#leet code= 106 
#time complexity =O(N)
#space complexity =O(N)
# Approach - postorder last value is the root. We will get the index using inoder. Anything to the left of inorder is left subtree, anything right of inoder list belongs to right subtree.



# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def buildTree(self, inorder, postorder):
        """
        :type inorder: List[int]
        :type postorder: List[int]
        :rtype: TreeNode
        """
        #edge case
        if len(inorder)==0: return None
        
        root=TreeNode(postorder[-1])
        index=inorder.index(postorder[-1])
        
        postleft=postorder[:index]
        postright=postorder[index:-1]
        inleft=inorder[:index]
        inright=inorder[index+1:]
        
        root.left=self.buildTree(inleft,postleft)
        root.right=self.buildTree(inright,postright)
        return root